src/modelfree/shaping_wrappers.py

```diff
@@ -0,0 +1,208 @@
+
+            if done[env_num]:
+                for rew_type in self.step_rew_dict:
+                    self.ep_rew_dict[rew_type].append(sum(self.step_rew_dict[rew_type][env_num]))
+                    self.step_rew_dict[rew_type][env_num] = []
+            self.counter += 1
+            self._log_sparse_dense_rewards()
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(BaseRLModel):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        Wrap an agent and add noise to its actions
+        :param agent: PolicyToModel(BaseRLModel) the agent to wrap
+        :param noise_annealer: Annealer.get_value - presumably the noise should be decreased
+        over time in order to get the adversarial policy to perform well on a normal victim.
+        :param noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.sess = agent.sess
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    def _get_noise_generator(self, noise_type):
+        noise_generators = {
+            'gaussian': lambda x, size: np.random.normal(scale=x, size=size)
+        }
+        return noise_generators[noise_type]
+
+    def predict(self, observation, state=None, mask=None, deterministic=False):
+        original_actions, states = self.agent.predict(observation, state, mask, deterministic)
+        action_shape = original_actions.shape
+
+        noise_param = self.noise_annealer()
+        noise = self.noise_generator(noise_param, action_shape)
+        normed_noise = original_actions * noise
+
+        noisy_actions = original_actions + normed_noise
+        return noisy_actions, states
+
+    def reset(self):
+        return self.agent.reset()
+
+    def setup_model(self):
+        pass
+
+    def learn(self):
+        raise NotImplementedError()
+
+    def action_probability(self, observation, state=None, mask=None, actions=None):
+        raise NotImplementedError()
+
+    def save(self, save_path):
+        raise NotImplementedError()
+
+    def load(self):
+        raise NotImplementedError()
+
+
+class HumanoidEnvWrapper(VecEnvWrapper):
+    def __init__(self, env):
+        super().__init__(env)
+        self.env = env
+
+    def reset(self):
+        return self.env.reset()
+
+    def step_wait(self):
+        obs, rew, done, infos = self.env.step_wait()
+        num_envs = len(obs)
+        for env_num in range(num_envs):
+            rew[env_num] -= infos[env_num]['reward_linvel']
```

Comment:
Can we not use `RewardShapingEnv` for this? I guess that assumes the info dicts have an axis corresponding to number of agents, but doesn't seem too hard to generalize to handle both? (Can have `agent_idx` be None (single-agent info dict), `0` or `1`.)